### ISSUE TITLE
TripletSemiHardLoss Returns NaN With Large Number of Classes

### ISSUE THREAD

**Issue by srilman** [tensorflow/addons#1676]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab
- TensorFlow version and how it was installed (source or binary): 2.2.0-r2
- TensorFlow-Addons version and how it was installed (source or binary): 0.9.1
- Python version: 3.6
- Is GPU used? (yes/no): yes

**Describe the bug**
Whenever I try using TripletSemiHardLoss in face-related applications (as it was originally designed for in the FaceNet paper), the loss value would always end up being NaN after 1 or 2 epochs. I wasn't sure why, since I never had that problem with the example in Tensorflow Addons repo (https://www.tensorflow.org/addons/tutorials/losses_triplet) which used MNIST. To try and replicate the issue, I tried changing the dataset used, and realized that datasets with around 100 or more classes cause this problem. This makes TripletSemiHardLoss really frustrating to use, since most face-related problems have thousands of classes.

**Code to reproduce the issue**
https://colab.research.google.com/github/tensorflow/addons/blob/master/docs/tutorials/losses_triplet.ipynb#scrollTo=TGBYNGxgVDrj
This example is just the Tensorflow Addons Triplet Loss tutorial (https://colab.research.google.com/github/tensorflow/addons/blob/master/docs/tutorials/losses_triplet.ipynb) that uses CIFAR100 instead of MNIST (and adjusts the model to take in the new input_shape) 

**Other info / logs**
Note that using TripletHardLoss doesn't cause this issue.
Also, I sometimes receive warnings regarding excessive tracing. This problem doesn't come up with the example code above.
```
...
WARNING:tensorflow:11 out of the last 11 calls to <function pairwise_distance at 0x7f9ad9903950> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/tutorials/customization/performance#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.
WARNING:tensorflow:11 out of the last 11 calls to <function triplet_semihard_loss at 0x7f9ad988e400> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/tutorials/customization/performance#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.
...
```

**Comment by gabrieldemarmiesse** [comment]
@lc0 could you take a look whenever you have the time? Thank you very much!

**Comment by srilman** [comment]
After a quick test playing with some of the params, I noticed that increasing the batch size decreases the likelihood of going to NaN. I think the problem is that with small batch sizes, it's very likely that some elements will have unique classes in the batch. Since triplet loss requires both positive and negative examples, there might be some kind of computation problem.

For example, using a batch size of 256 with CIFAR100 made it really unlikely for the loss to be NaN, since batch_size > num_classes.

**Comment by HauserA** [comment]
Triplet loss needs triplets to work (anchor image, positive image, negative image).
You need to make sure to always get enough triplets when you sample your batches.
If you sample randomly from your dataset, the number of triplets you get depends mostly on your batch_size and num_classes.

For batch_size=16 and num_classes=100 your expected value for the number of triplets is around 2.
For datasets with num_classes=2000 your expected values are:

1. batch_size = 16   --> 0.1 triplets
2. batch_size = 32   --> 0.5 triplets
3. batch_size = 64   --> 1.9 triplets
4. batch_size = 128 --> 7.8 triplets

Most of your images will not be used to compute the loss if you sample randomly.
For datasets with even more num_classes your chances to find triplets in your batches with random sampling will be close to 0.
If your batch has 0 triplets your loss will be NaN.

Try to sample the class ids and then at least 2 images per class id in the batch.
For example 16 unique person-ids with 4 images per person-id. This way all 64 images will be used to compute the loss (see:  [https://arxiv.org/abs/1903.07071](https://arxiv.org/abs/1903.07071)).

**Comment by srilman** [comment]
@HauserA That would make sense. I had started to realize that myself. Do you mind sharing how you calculated the expected number of triplets?

In any case I think it might help if the Triplet Loss example was modified so that this problem was a bit more apparent (maybe by selectively choosing elements for the batches as @HauserA describes).

**Comment by HauserA** [comment]
@srilman Yes, I made a colab notebook:
https://colab.research.google.com/drive/1qPjOIu5Dr9ADQmVfn7SXo-MPGSt8WT9x

I calculated the expected number with probability theory and just to be sure also with Montecarlo.

I also think it an example that shows how to selectively choose the elements with tf.data would be helpful. I split my dataset in hundreds of datasets for every class_id and use choose_from_datasets, but I think there may be a better way.

```
class TripletDataset(tf.data.Dataset):

    def _generator(images_per_id, nr_of_ids):
        for i in range(nr_of_ids): 
          yield ([i] * images_per_id)
    
    def __new__(cls, datasets, images_per_id):
        nr_of_ids = len(datasets)
        choice_dataset = tf.data.Dataset.from_generator( 
            cls._generator, 
            (tf.int64), 
            (tf.TensorShape([None])),
            args=(images_per_id, nr_of_ids)
        ) \
        .shuffle(nr_of_ids) \
        .unbatch()

        return tf.data.experimental.choose_from_datasets(datasets, choice_dataset)
```

**Comment by seanpmorgan** [comment]
Thanks for this helpful discussion! I'm going to close this, but if there is a concrete proposal to change the loss please feel free to open another issue. Would also be open to expanded tutorial

**Comment by AbrieJCronje** [comment]
@HauserA Would you mind sharing a colab notebook training a model with tensorflow-addons' tripletSemiHardLoss and the simplest version of your `class TripletDataset` on one of the `tensorflow-datasets` datasets? Any easily downloadable dataset would be great!

Here is example without using your `class TripletDataset` [https://colab.research.google.com/drive/1r04xVKqoSsItQFgXhOJfh-SLYDh-XRR0?usp=sharing](https://colab.research.google.com/drive/1r04xVKqoSsItQFgXhOJfh-SLYDh-XRR0?usp=sharing)
